### ISSUE TITLE
The result of the weights calculation in the build definitions making use of Intrinsics are different from the other builds. 

### ISSUE THREAD

**Issue by sfilipi** [dotnet/machinelearning#2998]
I am adding [a test](https://github.com/dotnet/machinelearning/pull/2048/files#r266634629) in [this PR](https://github.com/dotnet/machinelearning/pull/2048) and it is failing in the following definitions:

* core30 Build_Debug_Intrinsics
* core30 Build_Release_Intrinsics
* CentOS Build_Debug_Intrinsics
* Ubuntu Build_Release_Intrinsics

The weight calculation for MulticlassLR in those build definitions are different from the ones in the other definitions. 

weights in windows x64
[0, 0.5247429, -2.43042159, -0.490746975],
[0.210617781, 0, 0, -0.370147228],
[1.62008985E-06, -0.002228372, 2.49340868, 1.77681744]

weights in windows core30_debug_Intrinsics:
[0, 0.621862, -2.38398933, -0.427630424],
[0.108599916, 0, 0, -0.372884184],
[-0.106881194, -0.0351369865, 2.42675567, 1.80688608]

**Comment by sfilipi** [comment]
@Anipik  how can i run those builds locally? is there a flag to run with .net core 3.0 and intrinsics?

**Comment by Anipik** [comment]
Yes, git clean the repo and build the repo with flag build -debug-Intrinsics.
If you want to run tests using test explorer, changed the build configuration in the VS. (Dont Forget to restart VS)

**Comment by sfilipi** [comment]
After looking more into this, the difference yielded between the two types of implementations, for this [dataset and pipeline](https://github.com/dotnet/machinelearning/blob/3af9a5d96ade88e888894af23baef8fe4598f826/test/Microsoft.ML.Tests/TrainerEstimators/LbfgsTests.cs#L181) is comparable to the diverge we see in the weights calculated by varying the number of threads in the trainer. 

Outcome of the measurements for the weights calculations:

 **Multithred Debug Configuration**. Number of thread = 8  (This is the baseline)

| weights | 0           | 1            | 2          | 3            |
|---------|-------------|--------------|------------|--------------|
| 0       | 0           | 0.6630079    | -2.386228  | -0.468426347 |
| 1       | 0.217111193 | 0            | 0          | -0.329844654 |
| 2       | 3.59202E-06 | -0.084489964 | 2.41805434 | 1.79730487   |

 _**Single thread Debug Configuration**_

| weights | 0           | 1            | 2        | 3           |
|---------|-------------|--------------|----------|-------------|
| 0       | 0           | 0.7623239    | -2.34227 | -0.51741457 |
| 1       | 0.07323902  | 0            | 0        | -0.301424   |
| 2       | -0.09961926 | -0.143525153 | 2.37595  | 1.785325    |

Stats about the differences of weights with the **Multithred Debug Configuration**

| Max     | 0.143872173 |
|---------|-------------|
| Average | 0.054212062 |
  
**_Multithread Debug-Intrinsics Configuration._**. Number of threads = 8.

| weights | 0           | 1            | 2        | 3           |
|---------|-------------|--------------|----------|-------------|
| 0       | 0           | 0.7623239    | -2.34227 | -0.51741457 |
| 1       | 0.07323902  | 0            | 0        | -0.301424   |
| 2       | -0.09961926 | -0.143525153 | 2.37595  | 1.785325    |

Stats about the differences of weights with the **Multithred Debug Configuration**

| Max     | 0.138265    |
|---------|-------------|
| Average | 0.038507767 |

**_Single thread Debug-Intrinsics Configuration._**.

| weights | 0           | 1           | 2           | 3            |
|---------|-------------|-------------|-------------|--------------|
| 0       | 0           | 0.472711831 | -2.44474339 | -0.461801231 |
| 1       | 0.238091767 | 0           | 0           | -0.335685045 |
| 2       | 0           | 0           | 2.53461838  | 1.78326273   |

Stats about the differences of weights with the **Multithred Debug Configuration**

| Max     | 0.190296069 |
|---------|-------------|
| Average | 0.052316625 |

Stats about the differences of weights between  _**Single thread Debug Configuration**_ and **_Single thread Debug-Intrinsics Configuration._**.
| Max     | 0.289612069 |
|---------|-------------|
| Average | 0.106527889 |

**Comment by rosebyte** [comment]
Exact trained weights are not guaranteed across instruction sets, build configurations or thread counts because floating-point operations may be evaluated in a different order.
